File: src/partition-sample/src/main/scala/PartitionSample.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import org.apache.spark.ml.param._
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.util.{DefaultParamsReadable, Identifiable}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types._

/** Constants for <code>PartitionSample</code>. */
object PSConstants {
    final val ModeRS = "RandomSample"
    final val ModeHead = "Head"
    final val ModeATP = "AssignToPartition"

    final val rsAbsolute = "Absolute"
    final val rsPercent = "Percentage"

    final val newColDefault = "Partition"
}

trait PartitionSampleParams extends MMLParams {

  // TODO: Convert to Enum
  /** Sampling mode. The options are:
    * - AssignToPartition
    *   - seed
    *   - numParts
    *   - newColName
    * - RandomSample
    *   - mode: Absolute or Percentage
    *     - count (when Absolute)
    *     - percent (when Percentage)
    *   - seed
    * - Head
    *   - count
    *
    * The default setting is RandomSample
    * @group param
    */
  final val mode = StringParam(this, "mode", "AssignToPartition, RandomSample, or Head")
  setDefault(mode, PSConstants.ModeRS)
  /** @group getParam */
  final def getMode: String = $(mode)
  /** @group setParam */
  def setMode(value: String): this.type = set(mode, value)

  // TODO: Convert to Enum
  /** RandomSample mode
    * - Absolute
    * - Percentage
    *
    * Default is Percentage
    * @group param
    */
  final val rsMode = StringParam(this, "rsMode", "Absolute or Percentage", PSConstants.rsPercent)
  /** @group getParam */
  final def getRandomSampleMode: String = $(rsMode)
  /** @group setParam */
  def setRandomSampleMode(value: String): this.type = set(rsMode, value)

  // TODO: We need to create Option[Int] idiom for params
  /** Seed for random operations (RandomSplit or AssignToPartition
    * Default is -1
    * @group param
    */
  final val seed = LongParam(this, "seed", "seed for random ops", -1L)
  /** @group getParam */
  final def getSeed: Long = $(seed)
  /** @group setParam */
  def setSeed(value: Long): this.type = set(seed, value)

  /** Percentage of rows to return when Random Sampling
    * Default is .01
    * @group param
    */
  final val percent = DoubleParam(this, "percent", "percent of rows to return", 0.01)
  /** @group getParam */
  final def getPercent: Double = $(percent)
  /** @group setParam */
  def setPercent(value: Double): this.type = set(percent, value)

  /** Number of rows to return. Must be specified when the mode is HEAD, or when the mode is RandomSample
    * and the sampling mode is Absolute.
    *
    * Default is 1000
    * @group param
    */
  final val count = LongParam(this, "count", "number of rows to return", 1000L)
  /** @group getParam */
  final def getCount: Long = $(count)
  /** @group setParam */
  def setCount(value: Long): this.type = set(count, value)

  /** Name of the partition column. Specify a name when the mode is AssignToPartition
    *
    * Default is \"Partition\"
    * @group param
    */
  final val newColName = StringParam(this, "newColName", "name of the partition column", PSConstants.newColDefault)
  /** @group getParam */
  final def getNewColName: String = $(newColName)
  /** @group setParam */
  def setNewColName(value: String): this.type = set(newColName, value)

  /** Number of partitions when the mode is AssignToPartition
    *
    * Default is 10
    * @group param
    */
  final val numParts = IntParam(this, "numParts", "number of partitions", 10)
  /** @group getParam */
  final def getNumParts: Int = $(numParts)
  /** @group setParam */
  def setNumParts(value: Int): this.type = set(numParts, value)

  protected def validateAndTransformSchema(schema: StructType): StructType = {
    if (Seq(PSConstants.ModeHead, PSConstants.ModeRS).contains($(mode)))
      schema
    else
      ??? // schema + newCol
  }
}

object PartitionSample extends DefaultParamsReadable[PartitionSample]

// UID should be overridden by driver for controlled identification at the DAG level
/**
  * @param uid The id of the module
  */
sealed class PartitionSample(override val uid: String) extends Transformer with PartitionSampleParams {

  def this() = this(Identifiable.randomUID("PartitionSample"))

  override def transform(dataset: Dataset[_]): DataFrame = {
    $(mode) match {
      case PSConstants.ModeHead => dataset.limit(
        if ($(count) <= 2000000000) $(count).toInt else throw new Exception("Head limit 2b rows")).toDF
      case PSConstants.ModeRS => randomSample(dataset, $(rsMode), $(seed)).toDF
      case PSConstants.ModeATP => dataset.withColumn($(newColName), /* broken */ dataset.col("input"))
      case _ => ???
    }
  }

  def transformSchema(schema: StructType): StructType = {
    validateAndTransformSchema(schema)
  }

  def copy(extra: ParamMap): PartitionSample = defaultCopy(extra)

  private def randomSample(
    ds: Dataset[_],
    rsMode: String,
    seed: Long,
    replace: Boolean = false): Dataset[_] = {
        val frac = rsMode match {
            case PSConstants.rsPercent => $(percent)
            case PSConstants.rsAbsolute => $(count).toDouble / ds.count
            case _ => ???
        }
        println(s"Sampling ${ds.count} rows by ${frac * 100}% to get ~${ds.count * frac} rows")
        return ds.sample(replace, frac, seed)
  }

}

File: src/src/main/resources/mmlspark/PartitionSample.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class PartitionSample(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """
    Sampling mode.  The options are:

        - AssignToPartition
        - RandomSample
        - Head

    The default is RandomSample.

    Relevant parameters for the different modes are:

    - When the mode is AssignToPartition:
        - seed - the seed for random partition assignment.
        - numParts - the number of partitions.  The Default is 10.
        - newColName - the name of the partition column.
          The default is "Partition".
    - When the mode is RandomSample:
        - mode - Absolute or Percentage
        - count - the number of rows to assign to each partition when
          Absolute
        - percent - the percentage per partition when Percentage
    - When the mode is Head:
        - count - the number of rows

    Args:

        count (long): Number of rows to return (default: 1000)
        mode (str): AssignToPartition, RandomSample, or Head (default: RandomSample)
        newColName (str): Name of the partition column (default: Partition)
        numParts (int): Number of partitions (default: 10)
        percent (double): Percent of rows to return (default: 0.01)
        rsMode (str): Absolute or Percentage (default: Percentage)
        seed (long): Seed for random operations (default: -1)
    """

    @keyword_only
    def __init__(self, count=1000, mode="RandomSample", newColName="Partition", numParts=10, percent=0.01, rsMode="Percentage", seed=-1):
        super(PartitionSample, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.PartitionSample")
        self.count = Param(self, "count", "count: Number of rows to return (default: 1000)")
        self._setDefault(count=1000)
        self.mode = Param(self, "mode", "mode: AssignToPartition, RandomSample, or Head (default: RandomSample)")
        self._setDefault(mode="RandomSample")
        self.newColName = Param(self, "newColName", "newColName: Name of the partition column (default: Partition)")
        self._setDefault(newColName="Partition")
        self.numParts = Param(self, "numParts", "numParts: Number of partitions (default: 10)")
        self._setDefault(numParts=10)
        self.percent = Param(self, "percent", "percent: Percent of rows to return (default: 0.01)")
        self._setDefault(percent=0.01)
        self.rsMode = Param(self, "rsMode", "rsMode: Absolute or Percentage (default: Percentage)")
        self._setDefault(rsMode="Percentage")
        self.seed = Param(self, "seed", "seed: Seed for random operations (default: -1)")
        self._setDefault(seed=-1)
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, count=1000, mode="RandomSample", newColName="Partition", numParts=10, percent=0.01, rsMode="Percentage", seed=-1):
        """
        Set the (keyword only) parameters

        Args:

            count (long): Number of rows to return (default: 1000)
            mode (str): AssignToPartition, RandomSample, or Head (default: RandomSample)
            newColName (str): Name of the partition column (default: Partition)
            numParts (int): Number of partitions (default: 10)
            percent (double): Percent of rows to return (default: 0.01)
            rsMode (str): Absolute or Percentage (default: Percentage)
            seed (long): Seed for random operations (default: -1)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setCount(self, value):
        """

        Args:

            count (long): Number of rows to return (default: 1000)

        """
        self._set(count=value)
        return self


    def getCount(self):
        """

        Returns:

            long: Number of rows to return (default: 1000)
        """
        return self.getOrDefault(self.count)


    def setMode(self, value):
        """

        Args:

            mode (str): AssignToPartition, RandomSample, or Head (default: RandomSample)

        """
        self._set(mode=value)
        return self


    def getMode(self):
        """

        Returns:

            str: AssignToPartition, RandomSample, or Head (default: RandomSample)
        """
        return self.getOrDefault(self.mode)


    def setNewColName(self, value):
        """

        Args:

            newColName (str): Name of the partition column (default: Partition)

        """
        self._set(newColName=value)
        return self


    def getNewColName(self):
        """

        Returns:

            str: Name of the partition column (default: Partition)
        """
        return self.getOrDefault(self.newColName)


    def setNumParts(self, value):
        """

        Args:

            numParts (int): Number of partitions (default: 10)

        """
        self._set(numParts=value)
        return self


    def getNumParts(self):
        """

        Returns:

            int: Number of partitions (default: 10)
        """
        return self.getOrDefault(self.numParts)


    def setPercent(self, value):
        """

        Args:

            percent (double): Percent of rows to return (default: 0.01)

        """
        self._set(percent=value)
        return self


    def getPercent(self):
        """

        Returns:

            double: Percent of rows to return (default: 0.01)
        """
        return self.getOrDefault(self.percent)


    def setRsMode(self, value):
        """

        Args:

            rsMode (str): Absolute or Percentage (default: Percentage)

        """
        self._set(rsMode=value)
        return self


    def getRsMode(self):
        """

        Returns:

            str: Absolute or Percentage (default: Percentage)
        """
        return self.getOrDefault(self.rsMode)


    def setSeed(self, value):
        """

        Args:

            seed (long): Seed for random operations (default: -1)

        """
        self._set(seed=value)
        return self


    def getSeed(self):
        """

        Returns:

            long: Seed for random operations (default: -1)
        """
        return self.getOrDefault(self.seed)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.PartitionSample"

    @staticmethod
    def _from_java(java_stage):
        module_name=PartitionSample.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".PartitionSample"
        return from_java(java_stage, module_name)
